kernel/kernel-api/src/main/java/io/delta/kernel/Scan.java

```diff
@@ -217,6 +231,30 @@ public FilteredColumnarBatch next() {
                             "Column mapping mode is not yet supported: " + columnMappingMode);
                 }
 
+                // Type widening: add casts whenever the type of column read in the parquet file
+                // doesn't match the expected type from the table schema.
+                for (int colIdx = 0; colIdx < logicalReadSchema.length(); colIdx++) {
+                    StructField field = logicalReadSchema.at(colIdx);
+                    DataType parquetType =
+                            nextDataBatch.getSchema().get(field.getName()).getDataType();
+
+                    if (!parquetType.equals(field.getDataType())) {
+                        Cast cast =  new Cast(new Column(field.getName()), field.getDataType());
+
+                        // If creating the evaluators is expensive, we could consider caching and
+                        // reuse the same evaluator when the same cast in needed.
+                        ExpressionEvaluator evaluator =
+                            tableClient.getExpressionHandler().getEvaluator(
+                                nextDataBatch.getSchema(),
+                                cast,
+                                field.getDataType()
+                        );
+
+                        ColumnVector converted = evaluator.eval(nextDataBatch);
+                        nextDataBatch = nextDataBatch.withDeletedColumnAt(colIdx);
+                        nextDataBatch = nextDataBatch.withNewColumn(colIdx, field, converted);
```

Comment:
we are looking at adding methods on ExpressionHandler to be similar to `DataFrame`ish APIs. After that, evaluate once for each batch, instead of for every column.